(6.1) Update application access docs (#6055)
gravitational/teleport: docs/pages/application-access/guides/usage.mdx

everyone will log into Teleport.

In our example:

* `teleport.example.com` will host the Access Plane.
* `*.teleport.example.com` will host all of the applications e.g. `grafana.teleport.example.com`.

Teleport supports accessing these applications on other domains if required.
DNS entries must be configured and the correct certificates need to be obtained
for this to work.

```yaml
proxy_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3023
  tunnel_listen_addr: 0.0.0.0:3024
  public_addr: teleport.example.com:443
  # Example of two HTTPs keypairs.
  https_keypairs:
  - key_file: /etc/letsencrypt/live/teleport.example.com/privkey.pem
    cert_file: /etc/letsencrypt/live/teleport.example.com/fullchain.pem
  - key_file: /etc/letsencrypt/live/*.teleport.example.com/privkey.pem
    cert_file: /etc/letsencrypt/live/*.teleport.example.com/fullchain.pem
```

<Admonition
  type="tip"
  title="Using certbot to obtain wildcard certs"
>
  Let's Encrypt provides free wildcard certificates. If using [certbot](https://certbot.eff.org/)
  with DNS challenge, the below script will make setup easy. Update [EMAIL] and [DOMAIN]

>   ```sh

Reviewer: Let's replace this section with Teleport's ACME, I agree with Ben

  certbot certonly --manual \
    --preferred-challenges=dns \
    --email [EMAIL] \
    --agree-tos \
    --manual-public-ip-logging-ok \
    -d "teleport.example.com, *.teleport.example.com"
  ```
</Admonition>

> 

Reviewer 2: Can we replace this with ACME work and maybe add a sidebar on how it do if you already have your own certs.

Reviewer 3: Ideally we should keep both examples I think? ACME first and certbot for situations where people can't/don't want to use that.

Reviewer: It's OK to be opinionated in the docs, we will reduce the friction by giving customers our opinion. Folks who want certbot instead will know how to use it.

### Create a User

> In order to be able to access the application, a Teleport user needs to have

Reviewer: I would start with who needs what here: `A teleport user needs a role's permission to access an application.

a role that allows them to see it.

Teleport comes with a built-in `access` role that provides access to all
applications:

```sh

> $ tctl --config=/path/to/teleport.yaml users add --roles=access appuser

Reviewer: Props for using role preset, makes it so much easier

```

## Start Application Service with CLI Flags

A Teleport Application Proxy agent can be started with a single CLI command:

```sh
$ teleport start \
    --roles=app \
    --token=/tmp/token \
    --auth-server=teleport.example.com:3080 \
    --app-name="grafana" \
    --app-uri="http://localhost:3000" \
    --labels=env=dev
```

Note that the `--auth-server` flag must point to the Teleport cluster's proxy
endpoint because the application service always connects back to the cluster over
a reverse tunnel.

### Application Name

> When picking an application name, it's important to make sure the name will

Reviewer: I think if you use "Who can do what" it will be a bit easier to read. For example, users can access an application. Instead of an application can be accessed.

make a valid sub-domain (&lt;=63 characters, no spaces, only `a-z 0-9 _ -` allowed).

After Teleport is running, the application will be accessible at
`app-name.proxy_public_addr.com` e.g. `grafana.teleport.example.com`. Teleport
also provides the ability to override `public_addr` e.g `grafana.acme.com`
if you configure the appropriate DNS entry to point to the Teleport proxy server.

## Start Application Service with Config File

> Example `/etc/teleport-app.yaml` configuration:

Reviewer: where is this path coming from?

```yaml
teleport:
  # Data directory for the Application Proxy service. If running on the same
  # node as Auth/Proxy service, make sure to use different data directories.
  data_dir: /var/lib/teleport-app
  # Instructs the service to load the join token from the specified file
  # during initial registration with the cluster.
  auth_token: /tmp/token
  # Proxy address to connect to. Note that it has to be the proxy address
  # because the app service always connects to the cluster over a reverse
  # tunnel.
  auth_servers:
  - teleport.example.com:3080
app_service:
    enabled: yes
    # Teleport provides a small debug app that can be used to make sure application
    # access is working correctly. It'll output JWTs so it can be useful
    # when extending your application.
    debug_app: true
    # This section contains definitions of all applications proxied by this
    # service. It can contain multiple items.
    apps:
      # Name of the application. Used for identification purposes.
    - name: "grafana"
      # URI and port the application is available at.

>       uri: "http://localhost:3000"

Reviewer: This guide (or some other guide) should mention that if you have two apps with the same name on different hosts it will load balance them. (HA). May be for the future.

Author: Yeah I think dedicated HA guides for both app and db access are needed. Also, app access HA needs some work too, the way it's implemented right now it's not really HA because the session sticks to the app server for its entire duration.

      # Optional application public address to override.
      public_addr: "grafana.teleport.example.com"
      # Optional static labels to assign to the app. Used in RBAC.
      labels:
        env: "prod"
      # Optional dynamic labels to assign to the app. Used in RBAC.
      commands:
      - name: "os"
        command: ["/usr/bin/uname"]
        period: "5s"
auth_service:
  enabled: "no"
ssh_service:
  enabled: "no"
proxy_service:
  enabled: "no"
```

Start the application service:

```sh
$ teleport start --config=/etc/teleport-app.yaml
```

## Advanced Options

### Running Debug Application

For testing and debugging purposes we provide a built-in debug app. It can
be turned on using `debug_app: true`.

```yaml
app_service:
   enabled: yes
   debug_app: true
```

The app will dump all the request headers in the response.

### Customize Public Address

By default applications are available at `<app-name>.<proxy-host>:<proxy-port>`
address. To override the public address, specify the `public_addr` field:

```yaml
- name: "jira"
  uri: "https://localhost:8001"
  public_addr: "jira.example.com"
```

### Skip TLS Certificate Verification

> Teleport attempts to check if the certificates presented by the applications

Reviewer: I would add warning: danger zone because it's insecure

[Kubernetes Dashboard.](https://kubernetes.io/docs/tasks/access-application-cluster/web-ui-dashboard/).
The URI should be updated to include the subdirectory:

```yaml
- name: "k8s"
  uri: "http://10.0.1.60:8001/api/v1/namespaces/kube-system/services/https:kubernetes-dashboard:/proxy/#/overview?namespace=default"
  public_addr: "k8s.example.com"
```

### Rewrite

Application access supports simple rewrites. This is helpful for applications
that provide internal redirect headers:

```yaml
- name: "jira"
  uri: "https://localhost:8001"
  public_addr: "jira.example.com"
  rewrite:
    # Rewrite the "Location" header on redirect responses replacing the
    # host with the public address of this application.
    redirect:
    - "localhost"
    - "jenkins.internal.dev"
```

## View Applications in Teleport

Teleport provides a UI for quickly launching connected applications.

They can be viewed by navigating to a cluster and selecting Applications tab.
The URL structure looks like this:

`https://[cluster-url:cluster-port]/web/cluster/[cluster-name]/apps`

## Logging out of Applications

When you log into an application, you'll get a certificate and login session

> per your defined RBAC. If you want to force logout before this period you can
> force a log out by hitting the `/teleport-logout` endpoint:

Reviewer 3: ```suggestion
per your defined RBAC. If you want to force logout before this period you can
do so by hitting the `/teleport-logout` endpoint:
```
